snuba/datasets/metrics_aggregate_processor.py

```diff
@@ -0,0 +1,182 @@
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Mapping, Optional, Sequence, Tuple
+
+from snuba import environment, settings
+from snuba.consumers.types import KafkaMessageMetadata
+from snuba.processor import (
+    AggregateInsertBatch,
+    MessageProcessor,
+    ProcessedMessage,
+    _ensure_valid_date,
+)
+from snuba.query.expressions import (
+    Expression,
+    FunctionCall,
+    Literal,
+    OptionalScalarType,
+)
+from snuba.utils.metrics.wrapper import MetricsWrapper
+
+snuba_metrics = MetricsWrapper(environment.metrics, "metrics.processor")
+
+
+def _literal(value: OptionalScalarType) -> Literal:
+    return Literal(None, value)
+
+
+def _array_literal(values: Sequence[OptionalScalarType]) -> FunctionCall:
+    return FunctionCall(None, "array", tuple(map(_literal, values)))
+
+
+def _call(function_name: str, arguments: Tuple[Expression, ...]) -> FunctionCall:
+    return FunctionCall(None, function_name, arguments)
+
+
+METRICS_SET_TYPE = "s"
+METRICS_DISTRIBUTIONS_TYPE = "d"
+METRICS_COUNTERS_TYPE = "c"
+
+
+class MetricsAggregateProcessor(MessageProcessor, ABC):
+    TEN_SECONDS = 10
+    ONE_MINUTE = 60
+    ONE_HOUR = 3600
+    ONE_DAY = 3600 * 24
+    GRANULARITIES_SECONDS = [TEN_SECONDS, ONE_MINUTE, ONE_HOUR, ONE_DAY]
+
+    @abstractmethod
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    def process_message(
+        self, message: Mapping[str, Any], metadata: KafkaMessageMetadata
+    ) -> Optional[ProcessedMessage]:
+        if not self._should_process(message):
+            return None
+
+        timestamp = _ensure_valid_date(datetime.utcfromtimestamp(message["timestamp"]))
+        assert timestamp is not None
+
+        keys = []
+        values = []
+        tags = message["tags"]
+        assert isinstance(tags, Mapping)
+        for key, value in sorted(tags.items()):
+            assert key.isdigit()
+            keys.append(int(key))
+            assert isinstance(value, int)
+            values.append(value)
+
+        processed = [
+            {
+                "org_id": _literal(message["org_id"]),
+                "project_id": _literal(message["project_id"]),
+                "metric_id": _literal(message["metric_id"]),
+                "timestamp": _call(
+                    "toStartOfInterval",
+                    (
+                        _call("toDateTime", (_literal(timestamp.isoformat()),),),
+                        _call("toIntervalSecond", (_literal(granularity),)),
+                    ),
```

Comment:
Why don't you just round the date in the python code and just pass `toDateTime(THE_DATE_YOU_ROUNDED)`? 
That would save all the AST processing to create the expression and serialize it and make the whole thing more compact.

Reply:
I'm a bit hesitant to put in logic like https://stackoverflow.com/a/10854034 if we're trying to match the exact same behavior as `toStartOfInterval`. https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions/#tostartofintervaltime-or-data-interval-x-unit-time-zone is sort of unclear on details around interval boundaries. 

I think the key question is: are we going to receive queries looking for the window at a specific timestamp vs. querying within a range? I think these questions of "do you draw the day at the logical boundary vs. round down to `timestamp=(ms since epoch/86400 seconds day)`?" don't matter if we're expecting clients to query for ranges of timestamps, but if they query for `timestamp = {value}` then I think this isn't valuable enough to work on before a sanity test for the overall approach.

Reply:
> are we going to receive queries looking for the window at a specific timestamp vs. querying within a range

How we query does not matter. Since we are storing time buckets with multiple granularities in the same table, in order  for clickhouse to merge all the data points that represent a specific time bucket, all data points (rows we insert) need to be aligned with the beginning of the time bucket.

Let's say you insert three rows (granularity, one hour), one comes in with timestamp: 12:01, one 12:02, one 12:03, In order for clickhouse to merge all those three in one time bucket, they all must be written with timestamp = 12:00, otherwise they will not be merged in the same one hour bucket.

So since all rows in the table are aligned to the beginning of the time bucket they represent according to their granularity, there is only one way to query data: which is with time condition aligned with the beginning of the time bucket.

Reply:
> are we going to receive queries looking for the window at a specific timestamp vs. querying within a range? 

does this mean the difference for searching last 15min from a given timestamp versus inputing a range of two timestamps?

Reply:
Thanks for feedback, I think I talked myself into this being a bit more complicated than it is. Will update with in-process time rounding